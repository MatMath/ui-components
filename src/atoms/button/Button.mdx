---
name: Button
menu: Components
---

import { Playground, Props } from 'docz';
import Button from './Button';

# Button

A Button that allows users to take actions, and make choices, with a single tap. It can be a either:

- `Contained (filled) button`
- `Text button`

## Properties

<Props of={Button} />

## Usage

### Default variant for the button is `contained`

<Playground style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around' }}>
  <div>
    <Button type="danger">Danger Button</Button>
  </div>

<div>
  <Button type='dark'>Dark Button</Button>
</div>

<div>
  <Button type='light'>Light Button</Button>
</div>

<div>
  <Button type='secondary'>Secondary Button</Button>
</div>

  <div>
    <Button type="success">Success Button</Button>
  </div>
</Playground>

### You can have a text button by specifying `variant` property to `text`

<Playground style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around' }}>
  <div>
    <Button type="danger" variant="text">Danger Button</Button>
  </div>

<div>
  <Button type='dark' variant='text'>
    Dark Button
  </Button>
</div>

<div>
  <Button type='light' variant='text'>
    Light Button
  </Button>
</div>

<div>
  <Button type='secondary' variant='text'>
    Secondary Button
  </Button>
</div>

  <div>
    <Button type="success" variant="text">Success Button</Button>
  </div>
</Playground>

### In disabled mode, buttons will behave as following

<Playground className="test" style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around', flexFlow: 'wrap' }}>
<div style={{ display: 'flex', margin: '8px' }}>
  <div style={{margin: '8px'}}>
    <Button type="danger" disabled>Danger Button</Button>
  </div>

<div style={{ margin: '8px' }}>
  <Button type='dark' disabled>
    Dark Button
  </Button>
</div>

<div style={{ margin: '8px' }}>
  <Button type='light' disabled>
    Light Button
  </Button>
</div>

<div style={{ margin: '8px' }}>
  <Button type='secondary' disabled>
    Secondary Button
  </Button>
</div>

  <div style={{margin: '8px'}}>
    <Button type="success" disabled>Success Button</Button>
  </div>
</div>

<div style={{ display: 'flex', margin: '8px' }}>
  <div style={{margin: '8px'}}>
    <Button type="danger" variant="text" disabled>Danger Button</Button>
  </div>

<div style={{ margin: '8px' }}>
  <Button type='dark' variant='text' disabled>
    Dark Button
  </Button>
</div>

<div style={{ margin: '8px' }}>
  <Button type='light' variant='text' disabled>
    Light Button
  </Button>
</div>

<div style={{ margin: '8px' }}>
  <Button type='secondary' variant='text' disabled>
    Secondary Button
  </Button>
</div>

  <div style={{margin: '8px'}}>
    <Button type="success" variant="text" disabled>Success Button</Button>
  </div>
  </div>
</Playground>
