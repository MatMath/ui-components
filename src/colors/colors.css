:root {

/* -- DASH_GREEN -- */
--dash-green-00: #0E353D;
--dash-green-01: #3E5D64;
--dash-green-02: #6E868B;
--dash-green-03: #9FAEB1;
--dash-green-04: #CFD7D8;
--dash-green-05: #D9DFE0;
--dash-green-06: #F5F7F7;

/* -- DASH_DARKER_GREEN -- */
--dash-darker-green-00: #071A1E;

/* -- MID_GREEN -- */
--mid-green-00: #0E6476;
--mid-green-01: #3E8391;
--mid-green-02: #6EA2AD;
--mid-green-03: #9FC1C8;
--mid-green-04: #CFE0E4;
--mid-green-05: #D9E6E9;

/* -- PINK -- */
--pink-00: #FDB5AF;
--pink-01: #FDC4BF;
--pink-02: #FED3CF;
--pink-03: #FEE1DF;
--pink-04: #FFF0EF;
--pink-05: #FFF3F2;

/* -- ORANGE -- */
--orange-00: #FE5B33;
--orange-01: #FE7C5C;
--orange-02: #FE9D85;
--orange-03: #FFBDAD;
--orange-04: #FFDED6;
--orange-05: #FFE5DE;

/* -- FUNCTIONAL_RED -- */
--functional-red-00: #4E000D;
--functional-red-01: #8B0018;
--functional-red-02: #FF002D;
--functional-red-03: #FFA49F;

/* -- RED -- */
--red-00: #BF0A33;
--red-01: #CC3B5C;
--red-02: #D96C85;
--red-03: #E59DAD;
--red-04: #F2CED6;
--red-05: #F5D8DE;

/* -- GREY -- */
--grey-00: #615B57;
--grey-01: #999592;
--grey-02: #CCC7C2;
--grey-03: #D6D2CE;
--grey-04: #E0DDDA;
--grey-05: #EBE9E7;
--grey-06: #F5F4F3;

/* -- BLACK -- */
--black: #000000;

/* -- WHITE -- */
--white: #FFFFFF;

/* -- VALIDATOR_GREEN -- */
--validator-green: #20B422;

/* -- TRANSPARENT -- */
--transparent: transparent;
}
