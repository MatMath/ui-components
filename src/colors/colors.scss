/* -- DASH_GREEN -- */
$dash-green-00: #0e353d;
$dash-green-01: #3e5d64;
$dash-green-02: #6e868b;
$dash-green-03: #9faeb1;
$dash-green-04: #cfd7d8;
$dash-green-05: #d9dfe0;
$dash-green-06: #f5f7f7;

/* -- DASH_DARKER_GREEN -- */
$dash-darker-green-00: #071a1e;

/* -- MID_GREEN -- */
$mid-green-00: #0e6476;
$mid-green-01: #3e8391;
$mid-green-02: #6ea2ad;
$mid-green-03: #9fc1c8;
$mid-green-04: #cfe0e4;
$mid-green-05: #d9e6e9;

/* -- PINK -- */
$pink-00: #fdb5af;
$pink-01: #fdc4bf;
$pink-02: #fed3cf;
$pink-03: #fee1df;
$pink-04: #fff0ef;
$pink-05: #fff3f2;

/* -- ORANGE -- */
$orange-00: #fe5b33;
$orange-01: #fe7c5c;
$orange-02: #fe9d85;
$orange-03: #ffbdad;
$orange-04: #ffded6;
$orange-05: #ffe5de;

/* -- FUNCTIONAL_RED -- */
$functional-red-00: #4e000d;
$functional-red-01: #8b0018;
$functional-red-02: #ff002d;
$functional-red-03: #ffa49f;

/* -- RED -- */
$red-00: #bf0a33;
$red-01: #cc3b5c;
$red-02: #d96c85;
$red-03: #e59dad;
$red-04: #f2ced6;
$red-05: #f5d8de;

/* -- GREY -- */
$grey-00: #615b57;
$grey-01: #999592;
$grey-02: #ccc7c2;
$grey-03: #d6d2ce;
$grey-04: #e0ddda;
$grey-05: #ebe9e7;
$grey-06: #f5f4f3;

/* -- BLACK -- */
$black: #000;

/* -- WHITE -- */
$white: #fff;

/* -- VALIDATOR_GREEN -- */
$validator-green: #20b422;

/* -- ACCESSIBLE_VALIDATOR_GREEN -- */
$accessible-validator-green: #278637;

/* -- ACCESSIBLE_VALIDATOR_RED -- */
$accessible-validator-red: #e00635;

/* -- TRANSPARENT -- */
$transparent: transparent;
