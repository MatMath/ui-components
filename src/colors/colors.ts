/* -- DASH_GREEN -- */
export const dashGreen00 = '#0E353D';
export const dashGreen01 = '#3E5D64';
export const dashGreen02 = '#6E868B';
export const dashGreen03 = '#9FAEB1';
export const dashGreen04 = '#CFD7D8';
export const dashGreen05 = '#D9DFE0';
export const dashGreen06 = '#F5F7F7';

/* -- DASH_DARKER_GREEN -- */
export const dashDarkerGreen00 = '#071A1E';

/* -- MID_GREEN -- */
export const midGreen00 = '#0E6476';
export const midGreen01 = '#3E8391';
export const midGreen02 = '#6EA2AD';
export const midGreen03 = '#9FC1C8';
export const midGreen04 = '#CFE0E4';
export const midGreen05 = '#D9E6E9';

/* -- PINK -- */
export const pink00 = '#FDB5AF';
export const pink01 = '#FDC4BF';
export const pink02 = '#FED3CF';
export const pink03 = '#FEE1DF';
export const pink04 = '#FFF0EF';
export const pink05 = '#FFF3F2';

/* -- ORANGE -- */
export const orange00 = '#FE5B33';
export const orange01 = '#FE7C5C';
export const orange02 = '#FE9D85';
export const orange03 = '#FFBDAD';
export const orange04 = '#FFDED6';
export const orange05 = '#FFE5DE';

/* -- FUNCTIONAL_RED -- */
export const functionalRed00 = '#4E000D';
export const functionalRed01 = '#8B0018';
export const functionalRed02 = '#FF002D';
export const functionalRed03 = '#FFA49F';

/* -- RED -- */
export const red00 = '#BF0A33';
export const red01 = '#CC3B5C';
export const red02 = '#D96C85';
export const red03 = '#E59DAD';
export const red04 = '#F2CED6';
export const red05 = '#F5D8DE';

/* -- GREY -- */
export const grey00 = '#615B57';
export const grey01 = '#999592';
export const grey02 = '#CCC7C2';
export const grey03 = '#D6D2CE';
export const grey04 = '#E0DDDA';
export const grey05 = '#EBE9E7';
export const grey06 = '#F5F4F3';

/* -- BLACK -- */
export const black = '#000000';

/* -- WHITE -- */
export const white = '#FFFFFF';

/* -- VALIDATOR_GREEN -- */
export const validatorGreen = '#20B422';

/* -- ACCESSIBLE_VALIDATOR_GREEN -- */
export const accessibleValidatorGreen = '#278637';

/* -- ACCESSIBLE_VALIDATOR_RED -- */
export const accessibleValidatorRed = '#e00635';

/* -- TRANSPARENT -- */
export const transparent = 'transparent';
