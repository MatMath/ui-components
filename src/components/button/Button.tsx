import * as React from 'react';
import { getClassNames } from '@utility/cssUtils';
import styles from './Button.module.scss';

type ButtonNature = 'primary' | 'secondary' | 'danger' | 'ghost';
type ButtonSize = 'small' | 'medium' | 'large';
type ButtonTheme = 'light' | 'dark';

const defaultNature: ButtonNature = 'primary';
const defaultSize: ButtonSize = 'medium';
const defaultTheme: ButtonTheme = 'light';

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  nature?: ButtonNature;
  size?: ButtonSize;
  theme?: ButtonTheme;
}

const Button: React.FC<ButtonProps> = ({
  nature = defaultNature,
  size = defaultSize,
  theme = defaultTheme,
  children,
  ...htmlButtonProps
}) => {
  return (
    <button
      {...htmlButtonProps}
      className={getClassNames(
        styles.button,
        styles[nature],
        styles[size],
        styles[theme]
      )}
    >
      {children ? <span className={styles.content}>{children}</span> : null}
    </button>
  );
};

Button.defaultProps = {
  nature: defaultNature,
  size: defaultSize,
  theme: defaultTheme
};

export default Button;
