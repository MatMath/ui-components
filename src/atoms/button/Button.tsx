import * as React from 'react';
import { BaseButtonProps } from '@atoms/button/types';
import {
  ButtonDanger,
  ButtonDark,
  ButtonLight,
  ButtonSecondary,
  ButtonSuccess
} from '@ions';

export const ButtonTypes = {
  danger: ButtonDanger,
  secondary: ButtonSecondary,
  dark: ButtonDark,
  light: ButtonLight,
  success: ButtonSuccess
};

interface Props extends BaseButtonProps {
  type: keyof typeof ButtonTypes;
}

const Button: React.FunctionComponent<Props> = (props: Props): JSX.Element => {
  const renderButtonByType = ({ type, ...props }: Props): JSX.Element => {
    const ButtonName = ButtonTypes[type];
    if (!ButtonName) {
      throw new Error(
        'Unknown button type. Please refer to the documentation to see possible types'
      );
    }

    return <ButtonName {...props} />;
  };

  return renderButtonByType(props);
};

export default Button;
