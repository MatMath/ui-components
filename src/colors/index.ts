import { BasicColor, ColorExclude } from '@typings/Color.d';
import * as colors from '@design-tokens/colors.json';

export const $primary: ColorExclude<'200' | '400'> = colors.primary;

export const $secondary: BasicColor = colors.secondary;

export const $success: BasicColor = colors.success;

export const $danger: BasicColor = colors.danger;

export const $warning: BasicColor = colors.warning;

export const $info: BasicColor = colors.info;

export const $light: ColorExclude<'800'> = colors.light;

export const $transparent = 'transparent';
