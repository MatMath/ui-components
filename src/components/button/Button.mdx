---
name: Button
menu: Components
---

import { Playground, Props } from 'docz';
import { Button } from './Button';
import { CopyPasswordIcon } from '../../atoms/icons/CopyPasswordIcon';
import { white, midGreen00, dashGreen00 } from '../../colors/colors.ts';
import '../../utility/playgroundStyles.css';

# Button

A Button that allows users to take actions, and make choices, with a single tap.

## Properties

<Props of={Button} />

## Usage

### Nature

<Playground className={{preview: 'playground-base'}}>

    <Button nature='primary'>Primary</Button>
    <Button nature='secondary'>Secondary</Button>
    <Button nature='danger'>Danger</Button>
    <Button nature='ghost'>Ghost</Button>

</Playground>

### Possibilities

All buttons can display icons, text or mixed content.

<Playground className={{preview: 'playground-base'}}>
    <Button nature='primary'>
      <CopyPasswordIcon color={white} style={{ marginRight: '8px'}}/>
      Save
    </Button>

    <Button nature='secondary'>
      <CopyPasswordIcon color={midGreen00} style={{ marginRight: '8px'}}/>
      Cancel
    </Button>

    <Button nature='danger'>
      <CopyPasswordIcon color={white} style={{ marginRight: '8px'}}/>
      Delete
    </Button>

    <Button nature='ghost'>Decline</Button>

    <Button nature='ghost'>
      <CopyPasswordIcon color={midGreen00} />
    </Button>

</Playground>

### Theme

The default theme is `light` (white), but some buttons can be displayed in a `dark` theme (Dashlane's dash green, not an actual _dark_ theme).

<Playground className={{preview: 'playground-base'}} style={{ preview: {backgroundColor: dashGreen00} }}>
    <Button nature='primary' theme='dark'>
      Save
    </Button>

    <Button nature='secondary' theme='dark'>
      Cancel
    </Button>

    <Button nature='ghost' theme='dark'>
      Decline
    </Button>

    <Button nature='ghost' size='small' theme='dark'>
      <CopyPasswordIcon color={white} />
    </Button>

</Playground>

> ⚠️ `danger` nature is not supported in `dark` theme.

### Size

Buttons can be have different sizes [`small`, `medium`, `large`] with `medium` being the default value if none is specified.

<Playground className={{preview: 'playground-base'}} style={{ preview: {minHeight: '150px'} }}>
    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='primary' size='small'>Save</Button>
      <Button nature='primary' size='medium'>Save</Button>
      <Button nature='primary' size='large'>Save</Button>
    </div>

    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='secondary' size='small'>Cancel</Button>
      <Button nature='secondary' size='medium'>Cancel</Button>
      <Button nature='secondary' size='large'>Cancel</Button>
    </div>

    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='danger' size='small'>Delete</Button>
      <Button nature='danger' size='medium'>Delete</Button>
      <Button nature='danger' size='large'>Delete</Button>
    </div>

    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='ghost' size='small'>Decline</Button>
      <Button nature='ghost' size='medium'>Decline</Button>
      <Button nature='ghost' size='large'>Decline</Button>
    </div>

</Playground>

### Disabled

In disabled pseudo-state, buttons behave as following.

<Playground className={{preview: 'playground-base'}}>

    <Button nature='primary' disabled>Save</Button>
    <Button nature='secondary' disabled>Cancel</Button>
    <Button nature='danger' disabled>Delete</Button>
    <Button nature='ghost' size='medium' disabled>Decline</Button>
    <Button nature='ghost' disabled>
      <CopyPasswordIcon color={midGreen00}/>
    </Button>

</Playground>
