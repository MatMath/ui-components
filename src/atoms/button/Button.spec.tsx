import * as React from 'react';
import { mount, shallow } from 'enzyme';
import Button from './Button';
import { $danger, $light, $secondary, $success } from '@globals/colors';

describe('<Button>', () => {
  describe('Global render', () => {
    it('should render', () => {
      const ButtonWrapper = mount(<Button type='danger' />);

      expect(ButtonWrapper).toMatchSnapshot();
    });
  });

  describe('Themed button rendering', () => {
    it('should render a danger button', () => {
      const ButtonWrapper = mount(<Button type='danger'>Danger</Button>);
      expect(ButtonWrapper.find('BaseButton').prop('theme')).toMatchObject({
        background: $danger[500]
      });
    });

    it('should render a success button', () => {
      const ButtonWrapper = mount(<Button type='success'>Success</Button>);
      expect(ButtonWrapper.find('BaseButton').prop('theme')).toMatchObject({
        background: $success[500]
      });
    });

    it('should render a secondary button', () => {
      const ButtonWrapper = mount(<Button type='secondary'>Secondary</Button>);
      expect(ButtonWrapper.find('BaseButton').prop('theme')).toMatchObject({
        background: $secondary[500]
      });
    });

    it('should render a light button', () => {
      const ButtonWrapper = mount(<Button type='light'>Light</Button>);
      expect(ButtonWrapper.find('BaseButton').prop('theme')).toMatchObject({
        background: $light[300]
      });
    });

    it('should render a dark button', () => {
      const ButtonWrapper = mount(
        <Button type='dark' onClick={() => alert('hi')}>
          Dark
        </Button>
      );
      expect(ButtonWrapper.find('BaseButton').prop('theme')).toMatchObject({
        background: $light[400]
      });
    });
  });

  describe('No matching button type case', () => {
    it('should throw when the type is wrong', () => {
      expect(() =>
        // @ts-ignore
        shallow(<Button type='unknown'>Unknown</Button>)
      ).toThrowError(
        /Unknown button type. Please refer to the documentation to see possible types/
      );
    });
  });
});
