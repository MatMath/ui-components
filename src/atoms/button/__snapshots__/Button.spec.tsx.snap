// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Global render should render 1`] = `
<Button
  type="danger"
>
  <WrappedButton>
    <BaseButton
      theme={
        Object {
          "activeBackground": "#8b0018",
          "background": "#ff002d",
          "color": "#ffffff",
          "hoverBorder": "#8b0018",
        }
      }
    >
      <styled.button
        theme={
          Object {
            "activeBackground": "#8b0018",
            "background": "#ff002d",
            "color": "#ffffff",
            "hoverBorder": "#8b0018",
          }
        }
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": false,
                "lastClassName": "epmijA",
                "rules": Array [
                  "
  color: ",
                  [Function],
                  ";
  background-color: ",
                  [Function],
                  ";
  opacity: ",
                  [Function],
                  "
  cursor: ",
                  [Function],
                  "
  border: 1px solid ",
                  "transparent",
                  ";
  transition: all 0.2s ease-in;
  border-radius: 4px;  
  line-height: 1.75;
  padding: 6px 16px;
  min-width: 64px;
  font-family: 'Source Sans Pro', Helvetica, Arial, sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  outline: 0;
  &:hover, &:active {
    border: 1px solid ",
                  [Function],
                  "
  }
  &:active {
    background-color: ",
                  [Function],
                  ";
    border-color: ",
                  [Function],
                  "
    color: ",
                  [Function],
                  ";
  }
  &:disabled span {
    pointer-events: none;
  }
",
                ],
              },
              "displayName": "styled.button",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": "button",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          theme={
            Object {
              "activeBackground": "#8b0018",
              "background": "#ff002d",
              "color": "#ffffff",
              "hoverBorder": "#8b0018",
            }
          }
        >
          <button
            className="sc-htpNat epmijA"
          >
            <styled.span>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "EliYd",
                      "rules": Array [
                        "
  line-height: 1.75;
  font-family: 'Source Sans Pro', Helvetica, Arial, sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
",
                      ],
                    },
                    "displayName": "styled.span",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "span",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <span
                  className="sc-bwzfXH EliYd"
                >
                  Default
                </span>
              </StyledComponent>
            </styled.span>
          </button>
        </StyledComponent>
      </styled.button>
    </BaseButton>
  </WrappedButton>
</Button>
`;
